tests/functional/setups/manu_thermoporomech_nofrac_2d.py

```diff
@@ -0,0 +1,1219 @@
+        # AdArray.
+        external_sources = pp.ad.TimeDependentDenseArray(
+            name="source_energy",
+            domains=self.mdg.subdomains(),
+            previous_timestep=True,
+        )
+
+        # Add up contribution of internal and external sources of fluid
+        thermal_sources = internal_sources + external_sources
+        thermal_sources.set_name("Energy source")
+
+        return thermal_sources
+
+
+# -----> Solution strategy
+class ManuThermoPoroMechSolutionStrategy2d(
+    pp.thermoporomechanics.SolutionStrategyThermoporomechanics
+):
+    """Solution strategy for the verification setup."""
+
+    exact_sol: ManuThermoPoroMechExactSolution2d
+    """Exact solution object."""
+
+    fluid: pp.FluidConstants
+    """Object containing the fluid constants."""
+
+    results: list[ManuThermoPoroMechSaveData]
+    """List of SaveData objects."""
+
+    def __init__(self, params: dict):
+        """Constructor for the class."""
+        super().__init__(params)
+
+        self.exact_sol: ManuThermoPoroMechExactSolution2d
+        """Exact solution object."""
+
+        self.results: list[ManuThermoPoroMechSaveData] = []
+        """Results object that stores exact and approximated solutions and errors."""
+
+        self.flux_variable: str = "darcy_flux"
+        """Keyword to access the Darcy fluxes."""
+
+        self.stress_variable: str = "poroelastic_force"
+        """Keyword to access the poroelastic force."""
+
+    def set_materials(self):
+        """Set material parameters."""
+        super().set_materials()
+
+        # Instantiate exact solution object after materials have been set
+        self.exact_sol = ManuThermoPoroMechExactSolution2d(self)
+
+    def before_nonlinear_loop(self) -> None:
+        """Update values of external sources."""
+        super().before_nonlinear_loop()
+
+        sd = self.mdg.subdomains()[0]
+        data = self.mdg.subdomain_data(sd)
+        t = self.time_manager.time
+
+        # Mechanics source
+        mech_source = self.exact_sol.mechanics_source(sd=sd, time=t)
+        pp.set_solution_values(
+            name="source_mechanics", values=mech_source, data=data, time_step_index=0
+        )
+
+        # Flow source
+        flow_source = self.exact_sol.flow_source(sd=sd, time=t)
+        pp.set_solution_values(
+            name="source_flow", values=flow_source, data=data, time_step_index=0
+        )
+        # Energy source
+        energy_source = self.exact_sol.energy_source(sd=sd, time=t)
+        pp.set_solution_values(
+            name="source_energy", values=energy_source, data=data, time_step_index=0
+        )
+
+    def _is_nonlinear_problem(self) -> bool:
+        """The problem is non-linear due to the coupling between fluid flux and
+        advective energy transport.
+        """
+        return True
+
+    def bulk_modulus(self, subdomains: list[pp.Grid]) -> pp.ad.Operator:
```

Comment:
Does this belong to the SolutionStrategy, or should it be a constitutive law? Motivation: Being precise on classification/modularity may facilitate reuse in the future.

Reply:
It is rather specialized, so I think this placement will be fine.